### PR TITLE
MAILBOX-381 Vault implementation using MailRepository

### PR DESCRIPTION
**Pull request by chibenwa**
This PR is about implementing a Vault using mail repositories.

It is a subpart of https://github.com/linagora/james-project/pull/2191

It uses the mail converter supplied in https://github.com/linagora/james-project/pull/2194

### CODE REVIEW

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Override
+    public Publisher<Void> append(User user, DeletedMessage deletedMessage) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(deletedMessage);
+
+        MailRepository mailRepository = repositoryForUser(user);
+
+        return Mono.fromCallable(() -> mailRepository.store(mailConverter.toMail(deletedMessage)))
+            .publishOn(Schedulers.elastic())
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java]: publish as a last step looks weird, what are you trying to achieve?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        return Mono.fromRunnable(Throwing.runnable(() -> mailRepository.remove(mailKey)))
+            .publishOn(Schedulers.elastic())
+            .then();
+    }
+
+    @Override
+    public Publisher<DeletedMessage> search(User user, Query query) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(query);
+        Preconditions.checkArgument(query.getCriteria().isEmpty(), "Search is not supported yet...");
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java]: yet you can apply the query on the Flux, isn't it?

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java [hunk 3]
```diff
@@ -29,21 +29,42 @@
...
+        Preconditions.checkNotNull(path, "Supplied MailRepositoryPath valuie is null");
+
+        return new MailRepositoryPath(sanitizePath(path));
+    }
+
+    private static String sanitizePath(String path) {
+        if (!path.endsWith(PATH_DELIMITER)) {
+            return path;
+        }
+        return sanitizePath(path.substring(0, path.length() - 1));
```
Comment (mbaechler) [inline server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java]: is it a "strip last char" method?

**Comment by Arsnael** [comment]
Conflict with master, please rebase

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java (hunk 2 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java]: No. The ability to filter flux based on a query is done in #2193 and is not contributed yet.

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java (hunk 3 above)
Comment (chibenwa) [reply server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java]: Sounds like it yes

Comment (chibenwa) [reply server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java]:
```
    private static String sanitizePath(String path) {
        return StringUtils.stripEnd(path, "/");
    }
```

Makes it even more simple and avoids the recursion...

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java (hunk 2 above)
Comment (mbaechler) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java]: ok

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    @Override
+    public Publisher<Void> delete(User user, MessageId messageId) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+
+        MailRepository mailRepository = repositoryForUser(user);
+        MailKey mailKey = new MailKey(messageId.serialize());
+
+        return Mono.fromRunnable(Throwing.runnable(() -> mailRepository.remove(mailKey)))
+            .publishOn(Schedulers.elastic())
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java]: why?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        Preconditions.checkNotNull(query);
+        Preconditions.checkArgument(query.getCriteria().isEmpty(), "Search is not supported yet...");
+
+        MailRepository mailRepository = repositoryForUser(user);
+
+        try {
+            return Iterators.toFlux(mailRepository.list())
+                .publishOn(Schedulers.elastic())
+                .map(Throwing.function(mailRepository::retrieve))
+                .map(mailConverter::fromMail);
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java]: this step being CPU-intensive, you should switch to parallel, don't you think?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    @Override
+    public Publisher<Void> delete(User user, MessageId messageId) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+
+        MailRepository mailRepository = repositoryForUser(user);
+        MailKey mailKey = new MailKey(messageId.serialize());
+
+        return Mono.fromRunnable(Throwing.runnable(() -> mailRepository.remove(mailKey)))
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java]: don't you want the call to mailRepository to happen in elastic? If so, you should `subscribeOn`

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java [hunk 7]
```diff
@@ -25,25 +25,45 @@
...
 import com.google.common.base.Preconditions;
 
 public class MailRepositoryPath implements Comparable<MailRepositoryPath> {
-    public static final MailRepositoryPath fromEncoded(String encodedPath) throws UnsupportedEncodingException {
-        return new MailRepositoryPath(URLDecoder.decode(encodedPath, StandardCharsets.UTF_8.displayName()));
+
+    private static final String PATH_DELIMITER = "/";
+
+    public static MailRepositoryPath fromEncoded(String encodedPath) throws UnsupportedEncodingException {
+        Preconditions.checkNotNull(encodedPath, "Supplied MailRepositoryPath valuie is null");
```
Comment (mbaechler) [inline server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java]: s/valuie/value

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java [hunk 8]
```diff
@@ -25,25 +25,45 @@
...
+
+        return from(URLDecoder.decode(encodedPath, StandardCharsets.UTF_8.displayName()));
     }
 
-    public static final MailRepositoryPath from(String path) {
-        return new MailRepositoryPath(path);
+    public static MailRepositoryPath from(String path) {
+        Preconditions.checkNotNull(path, "Supplied MailRepositoryPath valuie is null");
```
Comment (mbaechler) [inline server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryPath.java]: s/valuie/value

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Override
+    public Publisher<Void> append(User user, DeletedMessage deletedMessage, InputStream inputStream) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(deletedMessage);
+        Preconditions.checkNotNull(inputStream);
+
+        MailRepository mailRepository = repositoryForUser(user);
+
+        return Mono.fromCallable(() -> mailRepository.store(mailConverter.toMail(deletedMessage, inputStream)))
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryDeletedMessageVault.java]: I didn't duplicate my comment about `subscribeOn` but it applies to all methods here I guess.

**Comment by Arsnael** [comment]
Read it

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**